pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -886,6 +898,95 @@ func (a *Actuator) requestPowerOn(ctx context.Context, baremetalhost *bmh.BareMe
 	return err
 }
 
+// isActionTimedOut checks if time since timestamp stored in specified machine annotation did not exceed given timeout
+func isActionTimedOut(machine *machinev1beta1.Machine, annotation string, timeout time.Duration) bool {
+	annotations := machine.Annotations
+	if annotations == nil {
+		return false
+	}
+	startTimeString, exist := annotations[annotation]
+	if !exist {
+		log.Printf("Annotation %q not found on machine %q", annotation, machine.Name)
+		return false
+	}
+	startTime, err := time.Parse(annotationTimestampFormat, startTimeString)
+	if err != nil {
+		log.Printf("Unable to parse time from annotation %q on machine %q", annotation, machine.Name)
+		return false
+	}
+	return time.Since(startTime) > timeout
+}
+
+// canReprovision checks if machine can be reprovisioned.
+// Machine can be reprovisione only if all of these conditions are met:
+//  * baremetalhost is not externally provisioned
+//  * machine is owned by a controller
+//  * machine role is not master
+func canReprovision(machine *machinev1beta1.Machine, baremetalhost *bmh.BareMetalHost) bool {
+	if baremetalhost.Spec.ExternallyProvisioned {
+		log.Printf("Reprovisioning of machine %q not allowed: BMH %q is externally provisioned", machine.Name, baremetalhost.Name)
+		return false
+	}
+	if metav1.GetControllerOf(machine) == nil {
+		log.Printf("Reprovisioning of machine %q not allowed: no owning controller", machine.Name)
+		return false
+	}
+	if machine.Labels[machineRoleLabel] == machineRoleMaster {
+		log.Printf("Reprovisioning of machine %q not allowed: has master role", machine.Name)
+		return false
+	}
+	return true
+}
+
+// getMhcByMachine returns MachineHealthCheck object responsible for given machine based on its label selectors
+func (a *Actuator) getMhcByMachine(machine *machinev1beta1.Machine) *machinev1beta1.MachineHealthCheck {
+	mhcOptions := client.ListOptions{
+		// Machine and MHC has to be in the same namespace
+		Namespace: machine.GetNamespace(),
+	}
+
+	mhcList := &machinev1beta1.MachineHealthCheckList{}
+	if err := a.client.List(context.TODO(), mhcList, &mhcOptions); err != nil {
+		log.Printf("Unable to get MHC objects: %s", err)
+		return nil
+	}
+
+	for _, mhc := range mhcList.Items {
+		selector, err := metav1.LabelSelectorAsSelector(&mhc.Spec.Selector)
+		if err != nil {
+			log.Printf("Unable to get machine selector from MHC %q: %s", mhc.GetName(), err)
+			return nil
```

Comment:
If there is a list of items in mhcList, should we ignore this error and continue instead of returning?

Reply:
Agree, will change it